keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,955 @@
+been set there, it will fall back transparently to logging through a global
+logger instance tracked in `k8s.io/klogr`, i.e. log calls never need to check
+whether they have a valid logger.
+
+### User Stories
+
+#### Story 1
+
+kube-scheduler developer Joan [wants to know which
+pod](https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-675074671)
+and which scheduler predicate log messages are associated with.
+
+When kube-scheduler starts processing a pod, it creates a new logger with
+`logger.WithValue(“pod”, klogr.KObj(pod))` and passes that around. While
+iterating over predicates, another logger gets created with
+`logger.WithName(<predicate name>)` and then is used when invoking that
+predicate.
+
+#### Story 2
+
+Scheduler-plugins developer John wants to increase the verbosity of the
+scheduler while it processes a certain pod ([“per-flow additional
+log”](https://github.com/kubernetes-sigs/scheduler-plugins/pull/289)).
+
+He does that by using `logger.V(0)` as logger for important pods and
+`logger.V(2)` as logger for less important ones. Then when the scheduler’s
+verbosity threshold is `-v=1`, a log message emitted with `V(1).InfoS` through
+the updated logger will be printed for important pods and skipped for less
+important ones.
+
+#### Story 3
+
+Kubernetes contributor Patrick is working on a unit test with many different
+test cases. To minimize overall test runtime, he allows different test cases to
+execute in parallel with `t.Parallel()`. Unfortunately, the code under test
+suffers from a rare race condition that is triggered randomly while executing
+all tests, but never when running just one test case at a time or when single
+stepping through it.
+
+He therefore wants to enable logging so that `go test` shows detailed log
+output for a failed test case, and only for that test case. He wants to run it
+like that by default in the CI so that when the problem occurs, all information
+is immediately available. This is important because the error might not show up
+when trying the same test invocation locally.
+
+When everything works, he wants `go test` to hide the log output to avoid
+blowing up the size of the CI log files.
+
+For each inner test case he adds a `NewTestContext(t)` invocation and uses the
+returned context and logger for that test case.
+
+#### Story 4
+
+Client developer Carlos wants to use client-go in his application, but is [less
+interested in log messages from
+it](https://kubernetes.slack.com/archives/CG3518SFJ/p1634217685020400). He
+makes the log messages from client-go less verbose by creating a logger with
+`logger.V(1)` and passing that to the client-go code. Now a `logger.V(3).Info`
+call in client-go is the same as a `logger.V(4).Info` in the application and
+will not be shown for `-v=3`.
+
+
+### Risks and Mitigations
+
+#### Design
+
+One risk is that code uses an uninitialized `logr.Logger`, which would lead to
```

Comment:
It looks like your are talking about a few independent problems in this section. Use lists or subtitles to make it clear.

Reply:
Will do.